Test train utils

tordbb opened the pull request:
Created a test for all functions inside of train_utils.py.

Commit:
Unfinished test for file test_train_utils.py

Commit:
Merge branch 'master' of https://github.com/acntech/fruit-classifier into test_train_utils

Commit:
Now tests two functions for crashing

The functions get_image_paths and get_data_and_labels are tested for
crashing when run.

Commit:
Merge branch 'master' of https://github.com/acntech/fruit-classifier into test_train_utils

Commit:
fixed bug where full folders were not removed.

Commit:
Added more tests to the file.

Commit:
Added tests for entire train_utils.py

Wanted to use one function per thing to test, but that would complicate
things. Each function depends on outputs from other functions, and the
test functions operates independently from eachother. Therefore, I could
not let one test_function store its result in the class object (self)
for others to use. If I had to test all functions in separate
test-functions, i would first need to run all functions and generate all
outputs in the setup function and store them in the class object. Then I would have to run them again in separate functions, extracting what I need from the class.

Commit:
One test per function to test - runs func in setUp

All functions of train_utils are now run in setUp, and parameters
describing them are stored in the object. Then, each set of parameters
are verified in separate tests for each function.

Commit:
Fixed camelcase and simplified

loeiten commented on test/train/test_train_utils.py:
```diff
@@ -0,0 +1,112 @@
+
+    def test_train_model(self):
+        # Test train_model output
+        self.assertEqual(self.num_epochs, 25)
```
No need to train for 25 epochs as this will take time, 2 is more than enough

tordbb replied:
Thanks, I needed to take a closer look at that!

loeiten commented on test/train/test_train_utils.py:
```diff
@@ -0,0 +1,112 @@
+        self.model_was_built = model._built
+
+        # Run train_model
+        history = train_model(model,
```
The `setUp` is for each member function of this class.
As there are 6 member functions you will train the same model with the same data and parameters 6 different times.
I would suggest to move the training to `test_train_model`

tordbb replied:
But I will not be able to let the output from test_train_model(), and use it in test_outputs_were_right()? In that case I should also verify the outputs (the 6 tests) inside of test_train_model()?

tordbb replied:
I have now gathered the entire training, aswell as the testing of each output into one function.

loeiten commented on test/train/test_train_utils.py:
```diff
@@ -0,0 +1,112 @@
+
+
+class TestTrainUtils(unittest.TestCase):
+    def setUp(self):
```
Recall that the `setUp` is common for all member functions.
If a variable is not shared with at least two member functions it should be refactored to the member function needing it.

For example:
Everything under
`# Run get_model_input`
is only needed in the `test_get_model_input` function, and should thus be moved there.

loeiten reviewed (approved):
Looks good, just needs some refactoring

loeiten reviewed (changes requested):
Sorry, my fat fingers managed to click `Approve` while I meant clicking `Request changes`

loeiten committed:
Merge branch 'master' into test_train_utils

loeiten committed:
Enhancements to README

Commit:
Gathered all 6 tests and training in 1 test +fixes

To avoid setUp from running the same training 6 times, the training was
gathered in one single test function. And since we otherwise cannot
store the outputs of the training for other tests to verify, I also
moved all the output verification into that function. Finally, I reduced
from training 25 epochs to 1 epoch.

Commit:
Increased to 2 epochs. Changed verification

loeiten commented on test/train/test_train_utils.py:
```diff
@@ -0,0 +1,90 @@
+        num_image_paths = len(image_paths)
+        self.assertGreater(num_image_paths, 0)
+
+        # Run get_data_and_labels and verify outputs
```
I see that I could have explained better.
I liked the way that you divided in functions previously, my only concern was that `setUp` was full of unrelated stuff

loeiten reviewed (changes requested):
Good refactoring of `setUp`, but it's better to split in small functions. I'll refactor, push and merge

loeiten committed:
Fixes to 3rd review of #10
